--- sound/pci/hda/cs35l41_hda_property.c	2025-06-23 07:44:01.166624107 +0000
+++ sound/pci/hda/cs35l41_hda_property.c	2025-06-23 07:44:39.458624768 +0000
@@ -12,6 +12,51 @@
 #include "cs35l41_hda_property.h"
 #include <linux/spi/spi.h>
 
+/*
+ * Device 10280BEB (Dell XPS 9530) doesn't have _DSD at all. Moreover, pin that is typically
+ * used for `speaker_id` is missing. SPI's cs-gpios definitions are also missing.
+ */
+static int dell_fiorano_no_acpi(struct cs35l41_hda *cs35l41, struct device *physdev, int id,
+                                const char *hid)
+{
+  struct cs35l41_hw_cfg *hw_cfg = &cs35l41->hw_cfg;
+  struct spi_device *spi = to_spi_device(cs35l41->dev);
+
+  /*
+   * 10280BEB has a firmware bug, which wrongly enables clock divider for intel-lpss
+   * Resultant SPI clock is 100Mhz/32767=3051Hz, which leads to ~3 minute hang on boot/wake up
+   * Avoid initializing device if lpss was not patched/fixed UEFI was not installed
+   */
+  if (spi->max_speed_hz < CS35L41_SPI_MAX_FREQ) {
+    dev_err(cs35l41->dev, "SPI's max_speed_hz is capped at %u Hz, will not continue to avoid hanging\n",
+            spi->max_speed_hz);
+    return -EINVAL;
+  }
+
+  dev_info(cs35l41->dev, "Adding DSD properties for %s\n", cs35l41->acpi_subsystem_id);
+
+  /* check SPI address to assign the index */
+  cs35l41->index = id;
+  cs35l41->channel_index = 0;
+  /* 10280BEB is missing pin which is typically assigned to `spk-id-gpios` */
+  cs35l41->speaker_id = cs35l41_get_speaker_id(physdev, cs35l41->index, 2, -1);
+  cs35l41->reset_gpio = gpiod_get_index(physdev, NULL, 1, GPIOD_OUT_LOW);
+
+  hw_cfg->spk_pos = cs35l41->index ? 1 : 0;      // 0th L, 1st R
+  hw_cfg->bst_type = CS35L41_EXT_BOOST;
+  hw_cfg->gpio1.func = CS35l41_VSPK_SWITCH;
+  hw_cfg->gpio1.valid = true;
+  hw_cfg->gpio2.func = CS35L41_INTERRUPT;
+  hw_cfg->gpio2.valid = true;
+  hw_cfg->valid = true;
+
+  /* Add second cs-gpio here */
+  if (cs35l41->index)
+    spi->cs_gpiod = gpiod_get_index(physdev, NULL, 0, GPIOD_OUT_HIGH);
+
+  return 0;
+}
+
 #define MAX_AMPS 4
 
 struct cs35l41_config {
@@ -454,7 +499,7 @@
 	{ "CLSA0101", NULL, lenovo_legion_no_acpi },
 	{ "CSC3551", "10280B27", generic_dsd_config },
 	{ "CSC3551", "10280B28", generic_dsd_config },
-	{ "CSC3551", "10280BEB", generic_dsd_config },
+	{ "CSC3551", "10280BEB", dell_fiorano_no_acpi },
 	{ "CSC3551", "10280C4D", generic_dsd_config },
 	{ "CSC3551", "103C89C6", generic_dsd_config },
 	{ "CSC3551", "103C8A28", generic_dsd_config },
